feat(accounts): Login with saved encrypted account

kitchenSpoon opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

<!-- Thanks for submitting a pull request! -->
<!-- Please provide enough information so that others can review your pull request -->
<!-- For more information, see the `CONTRIBUTING` guide. -->

## Description
<!--- Describe your changes in detail -->
Instead of saving the private key in the wallet file, save the encrypted key instead. Also asked for a passphrase when login using wallet file now that we are storing the encrypted key instead of private key. This is also backwards compatibility with wallet file that saves account with unencrypted private key.

This does not solve all the issues in but should be a step the right direction.
https://github.com/nos/client/issues/27

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- Describe the current and new situation/behavior --->
<!--- If it fixes an open issue, please link to the issue here. -->
It is not safe to store private key in a file. Saving the encrypted key instead is safer.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Tested locally.

## Screenshots (if appropriate):
![screen shot 2018-05-12 at 12 16 02 am](https://user-images.githubusercontent.com/2775497/39928824-b6db3048-5579-11e8-8330-573acb9fdede.png)
![screen shot 2018-05-12 at 12 15 58 am](https://user-images.githubusercontent.com/2775497/39928825-b70ddee4-5579-11e8-9895-7b5bf4aa22f2.png)


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Chore (tests, refactors, and fixes)
- [x] New feature (adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] I have read the **CONTRIBUTING** document.

## Closing issues
<!-- Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if such). -->
Fixes #

mhuggins commented on src/components/Register/AccountDetails/SaveAccount/SaveAccount.js:
```diff
@@ -84,7 +84,7 @@ export default class SaveAccount extends React.Component {
     }
 
-    const newAccount = new wallet.Account({ ...account, label });
+    const newAccount = new wallet.Account({ ...account, key: account.encryptedKey, label });
```
I don't think we can change this and still have it considered a NEP6 wallet.  The [proposal](https://github.com/neo-project/proposals/blob/master/nep-6.mediawiki#account) states:

* `key` is the private key of the account in the NEP-2 format. This field can be null (for watch-only address or non-standard address).

An alternative to this would be to store encrypted keys to local storage once they are entered in the "Passphrase" tab, and allow those to appear in a dropdown in that tab after being used once.

kitchenSpoon replied:
Hmm ok I get what you are saying, I'll go with your alternative method. Although i feel like there may be some errors with neon.js though in Account.js, it seems like the export function is taking the `_encrypted` variable and exporting it as the variable `key`, should it pehaps export it as `encryptedKey`? Is it because the AccountLike object does not have a `encryptedKey` field?
```
/**
   * Export Account as a AccountLike object.
   * @return {AccountLike}
   */
  export () {
    let key = null
    if (this._privateKey && !this._encrypted) throw new Error('Encrypt private key first!')
    if (this._encrypted) key = this._encrypted
    return {
      address: this.address,
      label: this.label,
      isDefault: this.isDefault,
      lock: this.lock,
      key,
      contract: this.contract,
      extra: this.extra
    }
  }
```

mhuggins replied:
Thanks for calling that out.  Perhaps I am unclear on the what the NEP6 proposal is saying.  I pinged snowypowers on Discord to ask about this since he wrote this code.  I'm interested in what he has to say.

If it's true that we can just use an encrypted version here, then this seems okay to me.  But it sounds like we would need to account for each account in the wallet being either a private key or an encrypted key.

mhuggins replied:
@kitchenSpoon Alright, I got some insight from snowy.

> NEP6 is always encrypted key or null (for watch only addresses)

So it seems we should have behavior for saving the wallet that's more in line with your original change here.  (I haven't gone through the additional commits yet.)

mhuggins replied:
@kitchenSpoon It looks like we can use `wallet.isNEP2` to see if it's encrypted.  Found [here](http://cityofzion.io/neon-js/docs/en/api-wallet.html#verify).

kitchenSpoon replied:
@mhuggins thanks!, let me know how it goes with snowypowers when he replies you.

kitchenSpoon commented:
@mhuggins I've added back my original changes and also checks for whether its a private key or a encrypted key. I've added a description to let user know whether the account selected is using a private key or encrypted key so they are not confused with why sometimes the passphrase input field appears.

mhuggins commented on src/login/components/LoginFormWalletFile/LoginFormWalletFile.js:
```diff
@@ -47,22 +52,59 @@ export default class LoginFormWalletFile extends React.Component {
-        {map(this.props.accounts, (account, index) => (
-          <option value={account.encrypted} key={`account${index}`}>{account.label}</option>
-        ))}
-      </Select>
```
I think we should have this function just return the `<Select>` as it was before, and call `this.renderPassphraseInput()` and `this.renderDescription()` directly from `render`.

mhuggins commented on src/login/components/LoginFormWalletFile/LoginFormWalletFile.js:
```diff
@@ -47,22 +52,59 @@ export default class LoginFormWalletFile extends React.Component {
-        ))}
-      </Select>
+      <div>
+        <Select className={styles.accounts} value={encryptedWIF} onChange={this.handleSelect}>
```
This admittedly was missing before you worked on it, but would you mind adding `disabled={disabled}` (based upon `this.props.disabled`) in case the login action is processing?

mhuggins commented on src/login/components/LoginFormWalletFile/LoginFormWalletFile.js:
```diff
@@ -47,22 +52,59 @@ export default class LoginFormWalletFile extends React.Component {
+          label="Passphrase"
+          placeholder="Enter passphrase"
+          value={passphrase}
+          onChange={this.handleChangePassphrase}
```
We need `disabled={disabled}` on this component as well.

mhuggins commented on src/login/components/LoginFormWalletFile/LoginFormWalletFile.js:
```diff
@@ -47,22 +52,59 @@ export default class LoginFormWalletFile extends React.Component {
+    const { encryptedWIF, passphrase } = this.props;
+
+    let input;
+    if (encryptedWIF !== '' && !wallet.isPrivateKey(encryptedWIF)) {
```
💅 Can we drop the `let` and inverse the logic here to simply return early if needed?  i.e.:

```js
if (encryptedWIF === '' || wallet.isPrivateKey(encryptedWIF)) {
  return null;
}

return (
  <Input ... />
);
```

mhuggins commented on src/login/components/LoginFormWalletFile/LoginFormWalletFile.js:
```diff
@@ -47,22 +52,59 @@ export default class LoginFormWalletFile extends React.Component {
+    const { encryptedWIF } = this.props;
+
+    if (encryptedWIF === '') {
+      return '';
```
Prefer `return null` here.

mhuggins commented on src/login/components/LoginFormWalletFile/LoginFormWalletFile.js:
```diff
@@ -47,22 +52,59 @@ export default class LoginFormWalletFile extends React.Component {
+    if (wallet.isPrivateKey(encryptedWIF)) {
+      return 'Private key detected';
+    } else {
+      return 'Encrypted key detected, please type passphrase';
```
Can we add punctuation to the end of both of these?

mhuggins commented on src/login/components/LoginFormWalletFile/LoginFormWalletFile.js:
```diff
@@ -76,19 +118,31 @@ export default class LoginFormWalletFile extends React.Component {
   handleSubmit = (event) => {
     event.preventDefault();
-    this.props.onLogin({ wif: this.props.wif });
+    const { encryptedWIF, passphrase = '', onLogin } = this.props;
```
You shouldn't need to assign `passphrase = ''`.  If it's not a string here, there's a bug.

mhuggins commented on src/login/components/LoginFormWalletFile/LoginFormWalletFile.js:
```diff
@@ -76,19 +118,31 @@ export default class LoginFormWalletFile extends React.Component {
 
   handleSelect = (event) => {
-    this.props.setWIF(event.target.value);
+    this.props.setEncryptedWIF(event.target.value);
```
Let's also reset the passphrase when a new item is selected from the dropdown.

```js
this.props.setPassphrase('')
```

jeroenptrs commented:
Great job @kitchenSpoon !
